### PR TITLE
Add PreLN to fsmt module

### PR DESCRIPTION
**Pull request by jinmang2**
# What does this PR do?

Add pre-layer normalization module to FSMT module to match fairseq transformers.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?
@stas00

### CODE REVIEW

File: src/transformers/models/fsmt/convert_fsmt_original_pytorch_checkpoint_to_pytorch.py
Comment (stas00) [inline src/transformers/models/fsmt/convert_fsmt_original_pytorch_checkpoint_to_pytorch.py:191]:
```suggestion
        "encoder_normalize_before": args.get("encoder_normalize_before", False),
```
this and the one below entry will fail with the original fairseq, so we need to check the key exists and set to False by default. 

I think the above should do the trick, but please test.

File: src/transformers/models/fsmt/convert_fsmt_original_pytorch_checkpoint_to_pytorch.py
Comment (stas00) [inline src/transformers/models/fsmt/convert_fsmt_original_pytorch_checkpoint_to_pytorch.py:196]: same as above

File: src/transformers/models/fsmt/modeling_fsmt.py
Comment (stas00) [inline src/transformers/models/fsmt/modeling_fsmt.py:544]: for better readability would it be possible to continue with checking the `encoder_normalize_before` flag? As the currently suggested `if ...` doesn't tell the reader why this check is done, and why it may or may be defined.

and then you can drop this
```
        else:
            self.layer_norm = None
```
same for decoder.

File: src/transformers/models/fsmt/modeling_fsmt.py
Comment (stas00) [inline src/transformers/models/fsmt/modeling_fsmt.py:618]: this looks like an editing ommision, one line up (616) should be deleted then as you already have it under `if`

**Review by stas00 — CHANGES_REQUESTED** [review]
Thank you for working on this, @jinmang2 

Overall looks good. I left a few suggestions.

Would it be better to name the new options `encoder_pre_layernorm` to better match the known concept? `normalize_before` is sort of asking for a noun after it. same for decoder.

The next step is to add a functional test, by building upon one of the existing fsmt tests.

Also if there are pre-trained models we could port that use preLN - it'd be great for then doing a quality test, which would be tricky to do with a dummy random data.

**Comment by stas00** [comment]
cc: @patil-suraj - if you'd like to have another pair of eyes on this one and also this will now create a bit of a diversion from your unmerged PR from long time ago. Not sure what you would want to do about it. It has kind of fallen between cracks due to the performance regression.

**Comment by sgugger** [comment]
This is typically the type of changes we usually don't accept inside one model architecture as it doesn't pass the test for new models: can an existing checkpoint for FSMT be used with this new config argument and give sensible results? -> No

I think this should warrant a new model.

**Comment by patil-suraj** [comment]
This is how BART was initially designed and was refactored to align with the general philosophy of the library to have self-contained model files. cf #9343. So this should be a new model. 

Also there are other fairseq models in the library that use `PreLN` for ex mBART and M2M100. Could you maybe check if you can load your checkpoints with these models ?

**Comment by jinmang2** [comment]
First of all, thank you for leaving a comment about my PR!

I wanted to port pororo's translation and wsd model to transformers. When I tested it myself, I got the same result as fairseq's reasoning result when I modified it with the corresponding PR.
- https://github.com/kakaobrain/pororo

We will check the following items mentioned above and mention them before 2/28.
- mBART

**Comment by sgugger** [comment]
> Not arguing whether it should be a new model or not, but unless I'm missing something, why the answer to *can an existing checkpoint for FSMT be used with this new config argument and give sensible results?* is No? I don't see any reason why it shouldn't work and not give identical results with this PR? It won't have any of the new model config entries and will work as intended.

When using an existing pretrained checkpoint with the config option `encoder_normalize_before=True` or `decoder_normalize_before=True`, the pretrained model will load properly (as all weights match) but a user won't get decent results with the model.

We can write a perfect modular toolbox that work with a current model implementation without breaking backward compatibility and have identical results, with just adding many new options in the configuration. That does not mean it conforms to the philosophy or pass the test of "this new option can be used with existing checkpoints".

**Comment by stas00** [comment]
> When using an existing pretrained checkpoint with the config option `encoder_normalize_before=True` or `decoder_normalize_before=True`, the pretrained model will load properly (as all weights match) but a user won't get decent results with the model.

If you're talking about the existing checkpoints - they won't have these config flags set in their config and as long as the default value for the new config options is `False` they will get identical results, as essentially they will be running the same model as before this PR.

If the defaults for the new config options are set to True, then indeed it'd break.

**Comment by stas00** [comment]
So talked some more with Sylvain and we aren't using the same test, hence the different pictures. He was using defaults `True` for new configs, I was using defaults `False` - hence the difference in the outcome.

And for me the idea of when using a new arch or not comes down to:

1. the config is for minor settings, like beam width, activation checkpointing, etc. which sometimes may lead to slightly less than optimal outcome but as close as possible to the best result.
2. any new config that changes the control flow (enables/disables code/changes order) which would lead to a different numerical outcome on the modeling level requires a new arch.

In other words no new architectural features can be added once a given modeling code is unleashed into the world and at least one checkpoint started using it, and a new architecture needs to be created.

**Comment by jinmang2** [comment]
I did a force-pushed on my local branch 30 minutes ago because of my commit mistake. If this behavior is a problem, please let me know! I will close the PR and commit a new one.

**Comment by jinmang2** [comment]
> This is how BART was initially designed and was refactored to align with the general philosophy of the library to have self-contained model files. cf https://github.com/huggingface/transformers/pull/9343. So this should be a new model.

> Also there are other fairseq models in the library that use PreLN for ex mBART and M2M100. Could you maybe check if you can load your checkpoints with these models ?

@patil-suraj Because there is no `SinusoidalPositionalEmbedding` class in mBART script, missing key occurs separately from the convert script. Therefore, porting with the script seems impossible.

**Comment by jinmang2** [comment]
> Overall looks good. I left a few suggestions.

@stas00 I checked all your suggestions and fixed them all in 2982403 commit!

> The next step is to add a functional test, by building upon one of the existing fsmt tests.

Does the functional test need to modify `test_modeling_fsmt.py`? Or can I provide the test result to the PR to see if the `pre_layernorm` option works normally?

> Also if there are pre-trained models we could port that use preLN - it'd be great for then doing a quality test, which would be tricky to do with a dummy random data.

Among the models of the PORORO library developed by OSLO developer hyunwoong ko, some models such as translation and word sense disambiguation use the `normalize_before` option. Is it okay to test the result and reflect it in the convert script?

File: src/transformers/models/fsmt/convert_fsmt_original_pytorch_checkpoint_to_pytorch.py
Comment (jinmang2) [reply src/transformers/models/fsmt/convert_fsmt_original_pytorch_checkpoint_to_pytorch.py]: I agree with your suggestion :)

File: src/transformers/models/fsmt/convert_fsmt_original_pytorch_checkpoint_to_pytorch.py
Comment (jinmang2) [reply src/transformers/models/fsmt/convert_fsmt_original_pytorch_checkpoint_to_pytorch.py]: I fixed it the same as the encoder

File: src/transformers/models/fsmt/modeling_fsmt.py
Comment (jinmang2) [reply src/transformers/models/fsmt/modeling_fsmt.py]: I agree with your review!

The original source code has been changed to the current one as follows to reflect the review.

- before review
```python
class FSMTEncoder(nn.Module):

    def __init__(self, config: FSMTConfig, embed_tokens: nn.Embedding):
        ...
        if config.encoder_normalize_before:
            self.layer_norm = nn.LayerNorm(embed_dim)
        else:
            self.layer_norm = None
        ...

    def forward(self, ...):
        ...
        if self.layer_norm is not None:
            x = self.layer_norm(x)
```
- after review
```python
class FSMTEncoder(nn.Module):

    def __init__(self, config: FSMTConfig, embed_tokens: nn.Embedding):
        ...
        self.pre_layernorm = config.encoder_pre_layernorm
        if self.pre_layernorm:
            self.layer_norm = nn.LayerNorm(embed_dim)
        ...

    def forward(self, ...):
        ...
        if self.pre_layernorm:
            x = self.layer_norm(x)
```

File: src/transformers/models/fsmt/modeling_fsmt.py
Comment (jinmang2) [reply src/transformers/models/fsmt/modeling_fsmt.py]: Thanks for finding the editing ommision! I also edited that part :)

**Comment by patil-suraj** [comment]
> > This is how BART was initially designed and was refactored to align with the general philosophy of the library to have self-contained model files. cf #9343. So this should be a new model.
> 
> > Also there are other fairseq models in the library that use PreLN for ex mBART and M2M100. Could you maybe check if you can load your checkpoints with these models ?
> 
> @patil-suraj Because there is no `SinusoidalPositionalEmbedding` class in mBART script, missing key occurs separately from the convert script. Therefore, porting with the script seems impossible.

I see. Could you try with `M2M100` then ? M2M100 also uses `PreLN` and `SinusoidalPositionalEmbedding`.

**Comment by stas00** [comment]
> > Overall looks good. I left a few suggestions.
> 
> @stas00 I checked all your suggestions and fixed them all in [2982403](https://github.com/huggingface/transformers/commit/298240385f570cae0616cf502b36b8bd909336d7) commit!

Thank you.

> > The next step is to add a functional test, by building upon one of the existing fsmt tests.
> 
> Does the functional test need to modify `test_modeling_fsmt.py`? Or can I provide the test result to the PR to see if the `pre_layernorm` option works normally?

I'm not sure I understand your question. What I meant is this new code can't be merged w/o tests exercising it.

And, of course, I'm not trying to contradict the maintainers who requested a new architecture. But it'd need a new test and it'd be something new and not `test_modeling_fsmt.py`, but you can of course model it after.
 
> > Also if there are pre-trained models we could port that use preLN - it'd be great for then doing a quality test, which would be tricky to do with a dummy random data.
> 
> Among the models of the [PORORO library](https://github.com/kakaobrain/pororo) developed by OSLO developer hyunwoong ko, some models such as translation and word sense disambiguation use the `normalize_before` option. Is it okay to test the result and reflect it in the convert script?

if these models fit the architecture by all means that would be perfect. Except we would want the model to be on the https://huggingface.co/models hub first, so that the test suite could run reliably.

**Pull request closed without merging**
